discoverd/client/client.go

```diff
@@ -1,261 +1,192 @@
+		urls = "http://127.0.0.1:1111"
 	}
-	return nil
+	return Config{Endpoints: formatURLs(strings.Split(urls, ","))}
 }
 
-func (c *Client) Ping() error {
-	return c.c.Get("/ping", nil)
+func formatURLs(urls []string) []string {
+	formatted := make([]string, len(urls))
+	for _, u := range urls {
+		if !strings.HasPrefix(u, "http") {
+			u = "http://" + u
+		}
+		formatted = append(formatted, u)
+	}
+	return formatted
 }
 
-type LeaderType string
-
-const (
-	LeaderTypeManual LeaderType = "manual"
-	LeaderTypeOldest LeaderType = "oldest"
-)
-
-type ServiceConfig struct {
-	LeaderType LeaderType `json:"leader_type"`
+func (c *Client) updateLeader(url string) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	for i, s := range c.servers {
+		if s.Host == url {
+			c.leader = i
+		}
+	}
 }
 
-func (c *Client) AddService(name string, conf *ServiceConfig) error {
-	if conf == nil {
-		conf = &ServiceConfig{}
+func (c *Client) httpClient(url string) *httpclient.Client {
+	checkRedirect := func(req *http.Request, via []*http.Request) error {
+		if len(via) >= 10 {
+			return fmt.Errorf("too many redirects")
+		}
+		if len(via) > 0 {
+			for attr, val := range via[0].Header {
+				if _, ok := req.Header[attr]; !ok {
+					req.Header[attr] = val
+				}
+			}
+		}
+		c.updateLeader(req.Host)
+		return nil
 	}
-	if conf.LeaderType == "" {
-		conf.LeaderType = LeaderTypeOldest
+	return &httpclient.Client{
+		URL: url,
+		HTTP: &http.Client{
+			Transport:     &http.Transport{Dial: dialer.Retry.Dial},
+			CheckRedirect: checkRedirect,
+		},
 	}
-	return c.c.Put("/services/"+name, conf, nil)
-}
-
-func (c *Client) RemoveService(name string) error {
-	return c.c.Delete("/services/" + name)
 }
 
-func (c *Client) Service(name string) Service {
-	return newService(c, name)
-}
+func (c *Client) Send(method string, path string, in, out interface{}) (err error) {
+	var leaderReq bool
+	switch method {
+	case "PUT", "DEL", "POST":
+		leaderReq = true
+	}
 
-func IsNotFound(err error) bool {
-	return hh.IsObjectNotFoundError(err)
-}
+	c.mu.RLock()
+	leader := c.leader
+	pinned := c.pinned
+	pinUpdated := c.pinUpdated
+	c.mu.RUnlock()
 
-func (c *Client) Instances(service string, timeout time.Duration) ([]*Instance, error) {
-	s := c.Service(service)
-	instances, err := s.Instances()
-	if len(instances) > 0 || err != nil && !IsNotFound(err) {
-		return instances, err
+	// try direct writes directly to the leader to avoid redirect
+	if leaderReq {
+		pinned = leader
 	}
 
-	events := make(chan *Event)
-	stream, err := s.Watch(events)
-	if err != nil {
-		return nil, err
-	}
-	defer stream.Close()
-	// get any current instances
-outer:
-	for event := range events {
-		switch event.Kind {
-		case EventKindCurrent:
-			break outer
-		case EventKindUp:
-			instances = append(instances, event.Instance)
+	for i := pinned; i < len(c.servers)+pinned; i++ {
+		k := i % len(c.servers)
+		hc := c.servers[k]
+		err := hc.Send(method, path, in, out)
+		if err != nil {
+			continue
```

Comment:
Hmm, would the best approach here be to collect these into a slice and then return them as a big Error? I could probably create a bunch of custom errors like ErrNoHosts etc to handle the different ways we can blow up.